server.go

```diff
@@ -227,23 +224,33 @@ func (s *Server) goScanConnection(connection net.Conn) {
 	go s.scan(scanCloser, client, tlsPeer)
 }
 
+
 func (s *Server) scan(scanCloser *ScanCloser, client string, tlsPeer string) {
 	defer s.wait.Done()
 	defer scanCloser.closer.Close()
 	for {
 		if s.readTimeoutMilliseconds > 0 {
 			scanCloser.closer.SetReadDeadline(time.Now().Add(time.Duration(s.readTimeoutMilliseconds) * time.Millisecond))
 		}
-		if scanCloser.Scan() {
-			// check in case scan took forever
-			select {
-			case <-s.doneTcp:
+		// we do the scan in a go func so we can listen to it finishing, along with our stop chan
+		// at the same time
+		scanned := make(chan bool)
+		go func() { scanned <- scanCloser.Scan() }()
+		select {
+		case <-s.stop:
+			return
+		case b := <-scanned:
+			if b {
+				// check in case scan took forever
```

Comment:
Not checking if the scanned channel was closed, but if `scanCloser.Scan()` returned false (i.e., it got an io.EOF and was the end of the stream).

I think that's still needed.  Could be wrong though :).  we previously were doing `if scanCloser.Scan()` so I was thinking this was replicating it.